akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -317,6 +317,20 @@ private[akka] final case class Grouped[T](n: Int) extends PushPullStage[T, immut
 /**
  * INTERNAL API
  */
+
+private[akka] final case class Limit[T](n: Int) extends PushStage[T, T] {
+  private var left = n
+
+  override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
```

Comment:
Very nice!

I'd probably encode it like this to avoid decrementing when it already known to be exceeded.

~~~scala
override def onPush(elem: T, ctx: Context[T]): SyncDirective =
  if (left > 0) {
    left -= 1
    ctx.push(elem)
   } else ctx.fail(new StreamLimitReachedException(n))
~~~